### PR TITLE
mon,auth,cephadm: support auth key rotation

### PR DESCRIPTION
**Pull request by liewegas**
This adds a low-level orchestrator command ``ceph orch daemon rotate-key <name>``.  Presumably we'll eventually build some automatic rotation on top of this, but this PR implements just the low-level infrastructure.

TODO
- [x] handle case when daemon is down
- [x] add support for other daemon types

## Checklist
- [ ] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/pybind/mgr/cephadm/services/cephadmservice.py
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:261]: If `auth get` fails, the keyring at the end will look like

```
[entity]
key = None
```

cause `keyring.splitlines()` will be an empty list. Did I miss anyhting?

**Comment by liewegas** [comment]
jenkins test api

File: src/common/ceph_context.cc
Comment (rzarzynski) [inline src/common/ceph_context.cc:185]: How about not naming the parameter in the implementations that don't make use of it? `[[maybe_unused]]` would be an alternative.

File: src/tools/rbd_wnbd/wnbd_handler.h
Comment (rzarzynski) [inline src/tools/rbd_wnbd/wnbd_handler.h:59]: nit: indentation.

File: src/mon/AuthMonitor.cc
Comment (rzarzynski) [inline src/mon/AuthMonitor.cc:1551]: Might be worth checking whether these were the sole users of the method.

File: src/pybind/mgr/cephadm/services/cephadmservice.py
Comment (rzarzynski) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:252]: A bit confused here. For `auth get-or-create` above we still examine `err` but for `auth caps` and `auth get` below we switched to `ret`. Is mixing these two ways intentional?

File: src/osd/OSD.cc
Comment (rzarzynski) [inline src/osd/OSD.cc:2778]: TODO(crimson): implement this.

File: src/auth/KeyRing.cc
Comment (rzarzynski) [inline src/auth/KeyRing.cc:143]: OK, `stringify()` does the same thing.

**Comment by neha-ojha** [comment]
> @neha-ojha @jdurgin Should we merge this piece before quincy? It isn't a complete solution, but it is a foundational piece and I worry if we leave it for too long this may go stale. I think the auth patches need a review still, though!

@liewegas do you want to address the review comments before dev freeze?

File: src/pybind/mgr/cephadm/services/cephadmservice.py
Comment (liewegas) [reply src/pybind/mgr/cephadm/services/cephadmservice.py:252]: nope, fixing!

File: src/auth/cephx/CephxKeyServer.cc
Comment (rzarzynski) [reply src/auth/cephx/CephxKeyServer.cc:255]: `get_used_pending_keys()` can just return the map instead of taking it as an input-output parameter.

File: src/mon/AuthMonitor.cc
Comment (rzarzynski) [reply src/mon/AuthMonitor.cc:1551]: Nope, we can't remove it.

File: src/pybind/mgr/cephadm/services/cephadmservice.py
Comment (rzarzynski) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:262]: @adk3798: could you please take a look on this fixup?

Comment (adk3798) [reply src/pybind/mgr/cephadm/services/cephadmservice.py:262]: I think what you changed looks correct, checking err has been our standard way of checking these commands' success so far.

**Comment by ljflores** [comment]
There are some tests failing that look related to this PR, i.e. http://pulpito.front.sepia.ceph.com/yuriw-2022-03-10_01:04:51-rados-wip-yuri5-testing-2022-03-07-0958-distro-default-smithi/6728452/.

**Comment by ljflores** [comment]
jenkins test windows

**Comment by rzarzynski** [comment]
Laura is right, the problem is related:

```
$ bin/ceph auth add client.xx mon 'allow *' osd "allow *"
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2022-03-22T23:42:56.453+0100 7fb2a34ed700 -1 WARNING: all dangerous and experimental features are enabled.
2022-03-22T23:42:56.465+0100 7fb2a228b700 -1 WARNING: all dangerous and experimental features are enabled.
added key for client.xx
$ bin/ceph auth get client.xx
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2022-03-22T23:43:06.125+0100 7f6e25c73700 -1 WARNING: all dangerous and experimental features are enabled.
2022-03-22T23:43:06.133+0100 7f6e25c73700 -1 WARNING: all dangerous and experimental features are enabled.
[client.xx]
	key = AQDwUDpiXOUqKBAAe2Fdhz/W8kSWrTIjkeYGbg==
```

should be:

```
$ bin/ceph auth get client.xx 
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2022-03-22T23:36:57.508+0100 7f13d0010700 -1 WARNING: all dangerous and experimental features are enabled.
2022-03-22T23:36:57.512+0100 7f13d0010700 -1 WARNING: all dangerous and experimental features are enabled.
[client.xx]
	key = AQCnTjpiuvQiJxAAEnKtHQABE2C9kbxXsHbc7w==
	caps mon = "allow *"
	caps osd = "allow *"
exported keyring for client.xx
```

DNM till fixing.

**Comment by rzarzynski** [comment]
jenkins retest this please

**Comment by adk3798** [comment]
jenkins test api

**Comment by rzarzynski** [comment]
Another bug popped up in http://pulpito.front.sepia.ceph.com/yuriw-2022-04-02_14:54:16-orch-wip-yuri11-testing-2022-04-01-0842-distro-default-smithi/6773812/:

```
2022-04-02T21:32:49.555 INFO:teuthology.orchestra.run.smithi146.stderr:++ ceph auth get-key mgr.y
2022-04-02T21:32:49.910 INFO:teuthology.orchestra.run.smithi146.stderr:+NK=AQD4aUhiCSlkCxAANVQf8PPD9eg8Cri81MR4eA==
2022-04-02T21:32:49.910 INFO:teuthology.orchestra.run.smithi146.stderr:+ '['AQD4aUhiCSlkCxAANVQf8PPD9eg8Cri81MR4eA== ==AQD4aUhiCSlkCxAANVQf8PPD9eg8Cri81MR4eA== ']'
2022-04-02T21:32:49.910 INFO:teuthology.orchestra.run.smithi146.stderr:+ sleep 5
```

**Comment by rzarzynski** [comment]
https://github.com/ceph/ceph/pull/46318 is merged. Taking the DNM off!

Could we schedule another QA run?

File: src/pybind/mgr/cephadm/module.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/module.py:1969]: probably need to add cephfs-mirror to this list at this point I think

File: doc/cephadm/operations.rst
Comment (adk3798) [inline doc/cephadm/operations.rst:102]: looking at some of the other code, it seems this isn't supported at all for mon daemons? If so, we may want to note that in the docs

File: src/pybind/mgr/cephadm/services/cephadmservice.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:33]:
```suggestion
    if daemon_type in ['rgw', 'rbd-mirror', 'cephfs-mirror', 'nfs', 'iscsi', 'ingress']:
```
NIT, no reason to use double quotes for that one specific type

File: src/pybind/mgr/cephadm/services/cephadmservice.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:27]: we should maybe change this to `_get_auth_entity` or something. We already have another function called `get_auth_entity` (that's calling this one) within the CephadmDaemonDeploySpec class which could be confusing.

**Review by adk3798 — APPROVED** [review]
Reviewing just from the cephadm side here. Just a few minor comments. Nothing I would block merging over, but things that would be nice to clean up (that I would be willing to do in a follow up PR if it's easier).

**Comment by adk3798** [comment]
jenkins test make check

**Comment by rzarzynski** [comment]
With the cephadm fix [the `rotate-keys` task is green at Sepia](http://pulpito.front.sepia.ceph.com/rzarzynski-2022-08-19_08:22:50-orch-main-distro-default-smithi/). I think we can restarts the QA for this PR.

**Comment by ljflores** [comment]
@rzarzynski from the latest tests:

https://pulpito.ceph.com/yuriw-2022-08-19_16:43:28-rados-wip-yuri8-testing-2022-08-19-0725-distro-default-smithi/6980811
```
2022-08-19T17:22:17.477 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:594: test_auth:  grep caps
2022-08-19T17:22:17.478 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:594: test_auth:  grep osd
2022-08-19T17:22:18.037 INFO:tasks.workunit.client.0.smithi141.stdout:	caps osd = "allow *"
2022-08-19T17:22:18.038 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:595: test_auth:  ceph auth get-key client.xx
2022-08-19T17:22:18.595 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:596: test_auth:  ceph auth print-key client.xx
2022-08-19T17:22:19.152 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:597: test_auth:  ceph auth print_key client.xx
2022-08-19T17:22:19.705 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:598: test_auth:  ceph auth caps client.xx osd 'allow rw'
2022-08-19T17:22:20.267 INFO:tasks.workunit.client.0.smithi141.stderr:updated caps for client.xx
2022-08-19T17:22:20.278 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:599: test_auth:  expect_false sh
2022-08-19T17:22:20.278 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:35: expect_false:  set -x
2022-08-19T17:22:20.278 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:36: expect_false:  sh
2022-08-19T17:22:20.827 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:36: expect_false:  return 0
2022-08-19T17:22:20.828 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:600: test_auth:  ceph auth get client.xx
2022-08-19T17:22:20.828 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:600: test_auth:  grep osd
2022-08-19T17:22:20.828 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:600: test_auth:  grep 'allow rw'
2022-08-19T17:22:21.392 INFO:tasks.workunit.client.0.smithi141.stdout:AQDGxv9iOFBDHRAAeWMtT1cVXEBsb/89JRjTvQ==AQDGxv9iOFBDHRAAeWMtT1cVXEBsb/89JRjTvQ==AQDGxv9iOFBDHRAAeWMtT1cVXEBsb/89JRjTvQ==	caps osd = "allow rw"
2022-08-19T17:22:21.392 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:601: test_auth:  ceph auth caps client.xx mon 'allow command "osd tree"'
2022-08-19T17:22:21.944 INFO:tasks.workunit.client.0.smithi141.stderr:updated caps for client.xx
2022-08-19T17:22:21.955 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:602: test_auth:  ceph auth export
2022-08-19T17:22:21.955 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:602: test_auth:  grep client.xx
2022-08-19T17:22:22.515 INFO:tasks.workunit.client.0.smithi141.stdout:[client.xx]
2022-08-19T17:22:22.516 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:603: test_auth:  ceph auth export -o authfile
2022-08-19T17:22:23.067 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:604: test_auth:  ceph auth import -i authfile
2022-08-19T17:22:23.633 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:605: test_auth:  check_response 'imported keyring'
2022-08-19T17:22:23.633 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:103: check_response:  expected_string='imported keyring'
2022-08-19T17:22:23.633 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:104: check_response:  retcode=
2022-08-19T17:22:23.634 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:105: check_response:  expected_retcode=
2022-08-19T17:22:23.634 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:106: check_response:  '[' '' -a '!=' ']'
2022-08-19T17:22:23.635 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:111: check_response:  grep --quiet -- 'imported keyring' /tmp/cephtool.gqB/test_invalid.A32
2022-08-19T17:22:23.636 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:112: check_response:  echo 'Didn'\''t find imported keyring in output'
2022-08-19T17:22:23.636 INFO:tasks.workunit.client.0.smithi141.stderr:Didn't find imported keyring in output
2022-08-19T17:22:23.637 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:113: check_response:  cat /tmp/cephtool.gqB/test_invalid.A32
2022-08-19T17:22:23.639 DEBUG:teuthology.orchestra.run:got remote process result: 1
2022-08-19T17:22:23.640 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:114: check_response:  exit 1
2022-08-19T17:22:23.641 INFO:tasks.workunit.client.0.smithi141.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1: check_response:  rm -fr /tmp/cephtool.gqB
2022-08-19T17:22:23.641 INFO:tasks.workunit:Stopping ['cephtool'] on client.0...
2022-08-19T17:22:23.643 DEBUG:teuthology.orchestra.run.smithi141:> sudo rm -rf -- /home/ubuntu/cephtest/workunits.list.client.0 /home/ubuntu/cephtest/clone.client.0
2022-08-19T17:22:23.937 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
```

File: src/mon/AuthMonitor.cc
Comment (rzarzynski) [inline src/mon/AuthMonitor.cc:1407]: Need to drop this from tests as well.

**Comment by rzarzynski** [comment]
Dropped the "imported keyring" check as we intentionally dropped the message (in https://github.com/ceph/ceph/pull/43655/commits/9dc0b7a4a0fe7976be31192f86608ba99bc1ed68).

BTW, I'm sorry for the labeling burden. I've pushed the testing branch by accident.

**Comment by rzarzynski** [comment]
jenkins test make check

**Comment by rzarzynski** [comment]
jenkins test api

**Comment by ljflores** [comment]
Rados suite review: http://pulpito.front.sepia.ceph.com/?branch=wip-yuri8-testing-2022-08-22-0646

Failures:
    1. https://tracker.ceph.com/issues/57269
    2. https://tracker.ceph.com/issues/52321
    3. https://tracker.ceph.com/issues/57270
    4. https://tracker.ceph.com/issues/55853
    5. https://tracker.ceph.com/issues/45721
    6. https://tracker.ceph.com/issues/37660
    7. https://tracker.ceph.com/issues/57122
    8. https://tracker.ceph.com/issues/57165
    9. https://tracker.ceph.com/issues/57303
    10 https://tracker.ceph.com/issues/56574
    11. https://tracker.ceph.com/issues/55986
    12. https://tracker.ceph.com/issues/57332

Details:
    1. rook: unable to read URL "https://docs.projectcalico.org/manifests/tigera-operator.yaml" - Ceph - Orchestrator
    2. qa/tasks/rook times out: 'check osd count' reached maximum tries (90) after waiting for 900 seconds - Ceph - Orchestrator
    3. cephadm: RuntimeError: Failed command: apt-get update: E: The repository 'https://download.ceph.com/debian-octopus jammy Release' does not have a Release file. - Ceph - Orchestrator
    4. test_cls_rgw.sh: failures in 'cls_rgw.index_list' and 'cls_rgw.index_list_delimited` - Ceph - RGW
    5. CommandFailedError: Command failed (workunit test rados/test_python.sh) FAIL: test_rados.TestWatchNotify.test - Ceph - RADOS
    6. smithi195:'Failing rest of playbook due to missing NVMe card' - Infrastructure - Sepia
    7. test failure: rados:singleton-nomsgr librados_hello_world - Ceph - RADOS
    8. expected valgrind issues and found none - Ceph - RADOS
    9. rados/cephadm: Failed to fetch package version from https://shaman.ceph.com/api/search/?status=ready&project=ceph&flavor=default&distros=ubuntu%2F22.04%2Fx86_64&sha1=b34ca7d1c2becd6090874ccda56ef4cd8dc64bf7  - Ceph - Orchestrator
    10. rados/valgrind-leaks: cluster [WRN] Health check failed: 2 osds down (OSD_DOWN)" in cluster log - Ceph - RADOS
    11. cephadm: Test failure: test_cluster_set_reset_user_config (tasks.cephfs.test_nfs.TestNFS) - Ceph - Orchestrator
    12. centos 8.stream and rhel 8.6 jobs fail to install ceph-test package due to xmlstarlet dependency - Ceph

**Comment by ljflores** [comment]
jenkins retest this please

**Comment by ljflores** [comment]
jenkins test make check

**Comment by ljflores** [comment]
jenkins test make check

**Comment by yuriw** [comment]
jenkins test make check

**Comment by adk3798** [comment]
> @liewegas @ljflores this PR was tested and ready for merge as soon as it passed `make check` re: https://trello.com/c/6s76bhl0

@yuriw  the make check failure was legitimate. I rebased this and added a commit to fix the test

**Comment by rzarzynski** [comment]
```
The following tests FAILED:
	 13 - run-tox-mgr-dashboard-check (Failed)
```

**Comment by rzarzynski** [comment]
jenkins test make check

**Comment by adk3798** [comment]
> ```
> The following tests FAILED:
> 	 13 - run-tox-mgr-dashboard-check (Failed)
> ```

unrelated, I think

```
ERROR: could not install deps [-rrequirements.txt, -cconstraints.txt, -rrequirements-test.txt, -rrequirements-extra.txt]; v = InvocationError('/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/.tox/check/bin/python -m pip install -rrequirements.txt -cconstraints.txt -rrequirements-test.txt -rrequirements-extra.txt', 1)
```

**Comment by rzarzynski** [comment]
All essential checks are green now! @yuriw, are we ready to merge?

**Comment by tchaikov** [comment]
before merging this PR, might want to squash the fixup commits?

**Comment by rzarzynski** [comment]
@tchaikov: squashed! No rebase conflicts during that. I think we're ready to go.

**Comment by adk3798** [comment]
> @tchaikov: squashed! No rebase conflicts during that. I think we're ready to go.

who would actually be merging this one? @yuriw ?

**Comment by ljflores** [comment]
@rzarzynski @adk3798 since this was rebased, does it need to go through another round of testing? Or was @adk3798's change trival?

**Comment by ljflores** [comment]
@adk3798: https://shaman.ceph.com/builds/ceph/wip-lflores-testing-2022-09-13-2119/8a7c1b85de9efffff6869a36a04ff3dd1438c8ba/

**Comment by ljflores** [comment]
Since @yuriw's last round of testing, one more commit was added to this PR. The commit (https://github.com/ceph/ceph/pull/43655/commits/b2085f0853b9d487a2f0275cdf401df970d85159) does not affect RADOS. All we need is to validate a relevant orch suite test. @adk3798 will validate/approve this test based on the recent build pasted above ^

**Comment by adk3798** [comment]
> Since @yuriw's last round of testing, one more commit was added to this PR. The commit ([b2085f0](https://github.com/ceph/ceph/commit/b2085f0853b9d487a2f0275cdf401df970d85159)) does not affect RADOS. All we need is to validate a relevant orch suite test. @adk3798 will validate/approve this test based on the recent build pasted above ^

run: https://pulpito.ceph.com/adking-2022-09-13_23:23:09-orch:cephadm-wip-lflores-testing-2022-09-13-2119-distro-default-smithi/, will review once it's complete

**Comment by adk3798** [comment]
> > Since @yuriw's last round of testing, one more commit was added to this PR. The commit ([b2085f0](https://github.com/ceph/ceph/commit/b2085f0853b9d487a2f0275cdf401df970d85159)) does not affect RADOS. All we need is to validate a relevant orch suite test. @adk3798 will validate/approve this test based on the recent build pasted above ^
> 
> run: https://pulpito.ceph.com/adking-2022-09-13_23:23:09-orch:cephadm-wip-lflores-testing-2022-09-13-2119-distro-default-smithi/, will review once it's complete

run finished

failures tracked by

- https://tracker.ceph.com/issues/57287
- https://tracker.ceph.com/issues/55986

Nothing new or related to this PR. Looks good to merge from orch side

**Pull request merged**
